### PR TITLE
feat: Validate Collection memory config through adding `conflictsWith` validation

### LINKED ISSUE #803: [config] Validation should fail if MaxAlloc, AvailableMemory, and MaxMemoryPercentage are all set
**Is your feature request related to a problem? Please describe.**

Refinery currently allows setting MaxAlloc, AvailableMemory, and MaxMemoryPercentage all in the same config.  In this instance, AvailableMemory and MaxMemoryPercentage will be used.  This can lead to confusion because a more explicit field, MaxAlloc, is also set. 

**Describe the solution you'd like**

Refinery falis validation if all 3 are set

**Describe alternatives you've considered**

**Additional context**

We ran into this with the Refinery helm chart.  The chart sets AvailableMemory and MaxMemoryPercentage by default, but nothing stops a user from setting MaxAlloc in their values.yaml.  This makes it seem like MaxAlloc will be used, but unless they remove AvailableMemory and MaxMemoryPercentage the resulting config will have all 3 values and use AvailableMemory and MaxMemoryPercentage instead.

### PR DESCRIPTION
**Pull request by MasonLegere**
## Which problem is this PR solving?

Fixes https://github.com/honeycombio/refinery/issues/803 by ensuring that only both `AvailableMemory` and `MaxAlloc` cannot be set at the same. `MaxMemoryPercentage` may still be set regardless as it has a default value defined. 

While making this changes I noticed that when only `AvailableMemory` was defined I ran into a validation `requireWith` error. The validation only fails on the first pass before defaults are loaded in from the config struct. If the failure is intentional then we probably should remove the default -- only breaking change I could see is if someone was running with validations turned off explicitly. 

```bash
 $ docker-compose up
[+] Running 1/0
 ✔ Container refinery-refinery-1  Recreated                                                                   0.0s
Attaching to refinery-refinery-1
refinery-refinery-1  | 2023/07/24 07:57:40 maxprocs: Leaving GOMAXPROCS=6: CPU quota undefined
refinery-refinery-1  | Validation failed for config file /etc/refinery/config.yaml:
refinery-refinery-1  |   the group Collection includes AvailableMemory, which also requires MaxMemoryPercentage
refinery-refinery-1  |
refinery-refinery-1  |
refinery-refinery-1 exited with code 1
$ cat config.yaml 
General:
  ConfigurationVersion: 2
  MinRefineryVersion: v2.0


Collection:
  AvailableMemory: 2Gb

```

## Short description of the changes

- Adds `conflictsWith` validation type acting in an opposite way to `requireWith` 
- Modified `requireWith` to check whether or not the required field has a default value

Fixes #803.

### CODE REVIEW

File: config/metadata/configMeta.yaml [hunk 1]
```diff
@@ -935,7 +935,8 @@ groups:
           allocation and `MaxAlloc` will be used instead. If set, then this must
           be a memory size. 64-bit values are supported. Sizes with standard
           unit suffixes (`MB`, `GiB`, etc.) and Kubernetes units (`M`, `Gi`, 
-          etc.) are also supported.
+          etc.) are also supported. If set, `Collections.maxAlloc` must not be
```
Comment (kentquirk) [inline config/metadata/configMeta.yaml:938]:
```suggestion
          etc.) are also supported. If set, `Collections.MaxAlloc` must not be
```

File: config/validate.go [hunk 2]
```diff
@@ -368,10 +368,17 @@ func (m *Metadata) Validate(data map[string]any) []string {
 					// if the named key is specified then this one must be also
 					otherName := validation.Arg.(string)
```
Comment (kentquirk) [inline config/validate.go:369]: Suggest adding a line that says "this validation cannot be used if the field has a default value".

**Review by kentquirk — APPROVED** [review]
This is good, thanks. Minor tweaks, please.

File: config/metadata/configMeta.yaml (hunk 1 above)
Comment (MasonLegere) [reply config/metadata/configMeta.yaml:938]: Ops thanks, I also found that the `AcceptOnlyListedKeys` docs need to be changed as well as it currently has the same issue (using requiredWith along with a default).

**Pull request merged**